### PR TITLE
LibRegex+LibJS: Implement the unicodeSets regex proposal

### PR DESCRIPTION
**Pull request by alimpfard**
why would anyone yakbait some poor soul into doing this? how cruel! how soul-crushing!

TODO:
- ~~nesting nested charclasses is borked~~
- ~~unicode properties are doing who-knows-what because we're ignoring some state~~
- the `\q{}` nonsense is not implemented, probably won't be in this PR
- ~~there are some regressions~~

```
Duration:
     -38.82s

Summary:
    Diff Tests:
        +46 ✅    -46 ❌

Diff Tests:
    test/built-ins/RegExp/unicodeSets/generated/character-class-difference-character-class-escape.js                ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-class-difference-character-class.js                       ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-class-difference-character-property-escape.js             ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-class-difference-character.js                             ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-class-escape-difference-character-class-escape.js         ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-class-escape-difference-character-class.js                ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-class-escape-difference-character-property-escape.js      ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-class-escape-difference-character.js                      ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-class-escape-intersection-character-class-escape.js       ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-class-escape-intersection-character-class.js              ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-class-escape-intersection-character-property-escape.js    ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-class-escape-intersection-character.js                    ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-class-escape-union-character-class-escape.js              ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-class-escape-union-character-class.js                     ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-class-escape-union-character-property-escape.js           ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-class-escape-union-character.js                           ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-class-intersection-character-class-escape.js              ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-class-intersection-character-class.js                     ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-class-intersection-character-property-escape.js           ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-class-intersection-character.js                           ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-class-union-character-class-escape.js                     ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-class-union-character-class.js                            ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-class-union-character-property-escape.js                  ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-class-union-character.js                                  ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-difference-character-class-escape.js                      ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-difference-character-property-escape.js                   ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-difference-character.js                                   ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-intersection-character-class-escape.js                    ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-intersection-character-class.js                           ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-intersection-character-property-escape.js                 ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-intersection-character.js                                 ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-property-escape-difference-character-class-escape.js      ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-property-escape-difference-character-property-escape.js   ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-property-escape-difference-character.js                   ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-property-escape-intersection-character-class-escape.js    ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-property-escape-intersection-character-class.js           ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-property-escape-intersection-character-property-escape.js ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-property-escape-intersection-character.js                 ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-property-escape-union-character-class-escape.js           ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-property-escape-union-character-class.js                  ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-property-escape-union-character-property-escape.js        ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-property-escape-union-character.js                        ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-union-character-class-escape.js                           ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-union-character-class.js                                  ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-union-character-property-escape.js                        ❌ -> ✅
    test/built-ins/RegExp/unicodeSets/generated/character-union-character.js                                        ❌ -> ✅

```

### CODE REVIEW

**Review by linusg — COMMENTED** [review]
LGTM,

- First commit could be split up more if you feel like it, the general `ParseFlags` changes could easily be in their own commit
- `Check code unit count range when accessing by code unit count` needs some sort of explanation IMO

File: Userland/Libraries/LibJS/AST.cpp [hunk 1]
```diff
@@ -3508,14 +3510,27 @@ Completion TemplateLiteral::execute(Interpreter& interpreter, GlobalObject& glob
...
+        return Value { js_string(global_object.heap(), String::empty()) };
+
+    char* buffer;
+    auto impl = StringImpl::create_uninitialized(total_size, buffer);
+
+    size_t offset = 0;
+    for (auto& string : raw_strings) {
+        __builtin_memcpy(buffer + offset, string.characters(), string.length());
+        offset += string.length();
+    }
```
[the remark this answers is not in the document]
Comment (alimpfard) [reply Userland/Libraries/LibJS/AST.cpp:3530]: Ah, we discussed this in discord and I forgot to explain this stuff

I could also just drop these from this PR, it's not actually related or required

**Comment by alimpfard** [comment]
> `Check code unit count range when accessing by code unit count` needs some sort of explanation IMO

:thinking: the wording is a bit shit, but the commit message really just explains what it's for - checking the right range.
Not sure how to improve that tbh

**Review by linusg — APPROVED** [review]
That's much nicer to look at now!

> Not sure how to improve that tbh

Fair enough, I just felt like someone with no context (me) could use a bit of an explanation, but no worries :)

**Pull request merged**